component/prometheus/kubernetes/podmonitors/crdManager.go

```diff
@@ -0,0 +1,189 @@
+	"github.com/prometheus/prometheus/scrape"
+	"k8s.io/client-go/tools/cache"
+
+	v1 "github.com/prometheus-operator/prometheus-operator/pkg/apis/monitoring/v1"
+	"github.com/prometheus-operator/prometheus-operator/pkg/client/versioned"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+// crdManager is all of the fields required to run the component.
+// on update, this entire thing will be recreated and restarted
+type crdManager struct {
+	opts   component.Options
+	logger log.Logger
+	config *Arguments
+
+	discovery        *discovery.Manager
+	scraper          *scrape.Manager
+	cg               configGenerator
+	discoveryConfigs map[string]discovery.Configs
+	scrapeConfigs    map[string]*config.ScrapeConfig
+
+	mut sync.Mutex
+}
+
+func newManager(opts component.Options, logger log.Logger, cfg *Arguments) *crdManager {
+	return &crdManager{
+		opts:             opts,
+		logger:           logger,
+		config:           cfg,
+		discoveryConfigs: map[string]discovery.Configs{},
+		scrapeConfigs:    map[string]*config.ScrapeConfig{},
+	}
+}
+
+func (c *crdManager) apply() {
+	c.mut.Lock()
+	defer c.mut.Unlock()
+	err := c.discovery.ApplyConfig(c.discoveryConfigs)
+	if err != nil {
+		level.Error(c.logger).Log("msg", "error applying discovery configs", "err", err)
+	}
+	scs := []*config.ScrapeConfig{}
+	for _, sc := range c.scrapeConfigs {
+		scs = append(scs, sc)
+	}
+	err = c.scraper.ApplyConfig(&config.Config{
+		ScrapeConfigs: scs,
+	})
+	if err != nil {
+		level.Error(c.logger).Log("msg", "error applying scrape configs", "err", err)
+	}
+	level.Debug(c.logger).Log("msg", "scrape config was updated")
+}
+
+func (c *crdManager) clearConfigs(kind string, ns string, name string) {
+	c.mut.Lock()
+	defer c.mut.Unlock()
+	prefix := fmt.Sprintf("%s/%s/%s", kind, ns, name)
+	for k := range c.discoveryConfigs {
+		if strings.HasPrefix(k, prefix) {
+			delete(c.discoveryConfigs, k)
+			delete(c.scrapeConfigs, k)
+		}
+	}
+}
+
+func (c *crdManager) addPodMonitor(pm *v1.PodMonitor) {
+	for i, ep := range pm.Spec.PodMetricsEndpoints {
+		pmc, err := c.cg.generatePodMonitorConfig(pm, ep, i)
+		if err != nil {
+			level.Error(c.logger).Log("name", pm.Name, "err", err, "msg", "error generating scrapeconfig from podmonitor")
+			continue
+		}
+		c.mut.Lock()
+		c.discoveryConfigs[pmc.JobName] = pmc.ServiceDiscoveryConfigs
+		c.scrapeConfigs[pmc.JobName] = pmc
+		c.mut.Unlock()
+	}
+	c.apply()
+}
+func (c *crdManager) onAddPodMonitor(obj interface{}) {
+	pm := obj.(*v1.PodMonitor)
+	level.Info(c.logger).Log("msg", "found pod monitor", "name", pm.Name)
+	c.addPodMonitor(pm)
+}
+func (c *crdManager) onUpdatePodMonitor(oldObj, newObj interface{}) {
+	pm := oldObj.(*v1.PodMonitor)
+	c.clearConfigs("podMonitor", pm.Namespace, pm.Name)
+	c.addPodMonitor(newObj.(*v1.PodMonitor))
+}
+func (c *crdManager) onDeletePodMonitor(obj interface{}) {
+	pm := obj.(*v1.PodMonitor)
+	c.clearConfigs("podMonitor", pm.Namespace, pm.Name)
+	c.apply()
+}
```

Comment:
nit: These are declared before they're used. I like to suggest that methods in a file be declared in use order so the file is easier to read top-down, but as this is now I'm forced to jump around in my reading. 

I would generally expect that the `on*PodMonitor` methods are defined _after_ `run`.